parlai/core/torch_generator_agent.py

```diff
@@ -1698,4 +1822,19 @@ def select_paths(self, logprobs, prior_scores, current_length):
         # Convert back to logspace.
         scores = sprobs[hyp_ids, choices].log()
         best_scores = prior_scores.expand_as(scores) + scores
-        return (hyp_ids, tok_ids, best_scores)
+
+        # need to resort because masking out of part of distribution can affect ranks
+        tok_ranks = (
```

Comment:
Here we use the double argsorting trick to compute ranks. We cannot use `sinds` as the first argsort computation because after it is computed, some values may be masked to `0` in `sprobs` on line 1817. Thus, we must run both `argsort` computations. 

We want the highest probability to have the lowest rank, so we set `descending=True`.

`choices` only contains one value for NucleusSampling since there are no beams (obviously) and this code is not batched. Thus, we flatten it to get the dimensions we want.

The lowest rank should be `1` not `0`, so we `+ 1` (this is also true for the other decoding schemes).

Reply:
Ah right, they will always be lower prob than chosen tokens, so they won't affect their ranks. So we can do `tok_ranks = torch.arange(sprobs.size[1]).to(logprobs.device)[choices.view(-1)] + 1`